pyomo/common/unittest.py

```diff
@@ -547,3 +554,404 @@ def assertRaisesRegex(self, expected_exception, expected_regex, *args, **kwargs)
+                tname = os.path.basename(dir_) + '_' + name
+
+                suffix = None
+                # Look for txt and yml file names matching py file names. Add
+                # a test for any found
+                for suffix_ in ['.txt', '.yml']:
+                    if os.path.exists(os.path.join(dir_, name + suffix_)):
+                        suffix = suffix_
+                        break
+                if suffix is not None:
+                    tname = tname.replace('-', '_')
+                    tname = tname.replace('.', '_')
+
+                    # Create list of tuples with (test_name, test_file, baseline_file)
+                    test_tuples.append(
+                        (tname, test_file, os.path.join(dir_, name + suffix))
+                    )
+
+        # Ensure a deterministic test ordering
+        test_tuples.sort()
+        return test_tuples
+
+    @classmethod
+    def gather_tests(cls, test_dirs):
+        # Find all .sh files in the test directories
+        sh_test_tuples = cls._find_tests(test_dirs, '*.sh')
+
+        # Find all .py files in the test directories
+        py_test_tuples = cls._find_tests(test_dirs, '*.py')
+
+        # If there is both a .py and a .sh, defer to the sh
+        sh_files = set(map(operator.itemgetter(1), sh_test_tuples))
+        py_test_tuples = list(
+            filter(lambda t: t[1][:-3] + '.sh' not in sh_files, py_test_tuples)
+        )
+
+        return py_test_tuples, sh_test_tuples
+
+    def check_skip(self, name):
+        """
+        Return a boolean if the test should be skipped
+        """
+
+        if name in self.solver_dependencies:
+            solvers_ = self.solver_dependencies[name]
+            if not all([self.solver_available[i] for i in solvers_]):
+                # Skip the test because a solver is not available
+                _missing = []
+                for i in solvers_:
+                    if not self.solver_available[i]:
+                        _missing.append(i)
+                return "Solver%s %s %s not available" % (
+                    's' if len(_missing) > 1 else '',
+                    ", ".join(_missing),
+                    'are' if len(_missing) > 1 else 'is',
+                )
+
+        if name in self.package_dependencies:
+            packages_ = self.package_dependencies[name]
+            if not all([self.package_available[i] for i in packages_]):
+                # Skip the test because a package is not available
+                _missing = []
+                for i in packages_:
+                    if not self.package_available[i]:
+                        _missing.append(i)
+                return "Package%s %s %s not available" % (
+                    's' if len(_missing) > 1 else '',
+                    ", ".join(_missing),
+                    'are' if len(_missing) > 1 else 'is',
+                )
+
+            # This is a hack, xlrd dropped support for .xlsx files in 2.0.1 which
```

Comment:
Should we do this here, or should we do this in `setup.py`?

Reply:
I think here.  Neither pandas nor xlrd are required dependencies, and this is really just guarding an edge case that cropped up in testing dataportals.